### PR TITLE
Rearrangement Algorithm Implementation

### PR DESCRIPTION
**Pull request by ckcks12**
#1464 

From 
image

To
image

# Algorithm
Sort them by *Earlier -> Up, Longer -> Left.*
And then group them who are being overlapped each other.
And give number by order then stretch them to the same size except the last one. *last one could be greedy like he can eat all the rest place!*

# Performance

### original
image
### with new algorithm
image

### original
image
### with new algorithm
image

## About the Optimization

I leave some comments like **TODO** where further optimization would be effective.
But as you can see above Performance Metrics, even a thousand of events doesn't affect much.
in case you want to know how A THOUSAND of events looks like,
image

The source might be kinda dirty and hard to read because I hardly use the `map()` for loop which would be prettier. Honestly it causes a performance issue. Here you can find why I used `for`. 
https://github.com/dg92/Performance-Analysis-JS

# And...
I put some more like padding between events. like `2px` between them.
I think it's prettier but also worry bout it's hard coded.

### CODE REVIEW

File: src/DayColumn.js [hunk 1]
```diff
@@ -13,6 +13,18 @@ import * as DayEventLayout from './utils/DayEventLayout'
 import TimeSlotGroup from './TimeSlotGroup'
 import TimeGridEvent from './TimeGridEvent'
 
+function dfs(node, maxIdx, visited) {
```
Comment (ckcks12) [inline src/DayColumn.js]: I put it outside on purpose bc of hoisting. and also I think it's better than declared inside of class.

**Comment by KhaledMohamedP** [comment]
@ckcks12 Great job 👏 

This PR addressees #1440

**Comment by beaugunderson** [comment]
@ckcks12 this looks great; we're very excited about getting this merged!

**Comment by jquense** [comment]
thanks for doing this, hoping to get some time to review in the next few days!

**Comment by KhaledMohamedP** [comment]
Looks like a lot of people are waiting for this PR to be merge! @jquense, did you get chance to review this PR?

Appreciate your time and help here 🙏

**Comment by nikhilpingle** [comment]
Nice fix. I also need to have a couple of more enhancement in UI. I hope the community welcomes them as well.

File: src/DayColumn.js (hunk 1 above)
Comment (jquense) [inline src/DayColumn.js]: don't abbreviate the name plz

Comment (ckcks12) [reply src/DayColumn.js]: Okayy.

File: examples/events.js [hunk 2]
```diff
@@ -111,4 +111,52 @@ export default [
...
+    title: 'Cooking Class',
+    start: new Date(2015, 3, 14, 17, 30, 0),
+    end: new Date(2015, 3, 14, 19, 0, 0),
+  },
+  {
+    id: 23,
+    title: 'Go to the gym',
+    start: new Date(2015, 3, 14, 18, 30, 0),
+    end: new Date(2015, 3, 14, 20, 0, 0),
+  },
```
Comment (ckcks12) [inline examples/events.js]: These are for overlapping situation

File: src/TimeGridEvent.js [hunk 3]
```diff
@@ -42,10 +46,10 @@ function TimeGridEvent(props) {
...
         style={{
           ...userProps.style,
-          top: `${top}%`,
-          height: `${height}%`,
-          [rtl ? 'right' : 'left']: `${Math.max(0, xOffset)}%`,
-          width: `${width}%`,
+          top: stringifyPercent(top),
+          [rtl ? 'right' : 'left']: stringifyPercent(xOffset),
+          width: stringifyPercent(width),
+          height: stringifyPercent(height),
```
Comment (ckcks12) [inline src/TimeGridEvent.js]: Sometimes the algorithm needs `calc()` so I just made it as `'string' || 'number'`. number will be followed by `%`.

Maybe you think it is no good that I didn't make it all `string` only. Frankly I tried it but you know that plain css string is quite annoying for testing, right? 

image

I also tried to `parse()` and `floor()` but... yeah... too much. so this is why.

**Comment by ckcks12** [comment]
@jquense Hallo~

I made it as strategy pattern like pluggable design. :laughing: 

`DayEventLayout.js` -> `layout-algorithms/overlap.js` or `layout-algorithms/no-overlap.js`

And from the top of the library, the `Calendar`, I made new attribute `dayLayoutAlgorithm` and its prop type is `overlap` | `no-overlap` | `Function`.

and also I created a demo section for `no-overlap` algorithm.
image



Sooooooo hope you like it :eyes:

**Comment by jquense** [comment]
just need a moment to try it out then gtg

**Comment by KhaledMohamedP** [comment]
@jquense did you find any blockers?

**Comment by Canuckaholic** [comment]
Very excited for this! Thanks for all the work on this everyone involved.

**Comment by Dragomir-Ivanov** [comment]
Having overlap strategy as an option to Calendar would be very helpful. I also need all events visible at the cost of ugly presentation.

**Comment by KhaledMohamedP** [comment]
@jquense Thanks for merging :) 
 
Can we please push this to npm and increment the package version as a `MINOR` update? 

Looking forward to this update 😄

**Pull request merged**